hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketAssignFunction.java

```diff
@@ -194,9 +194,10 @@ private void processRecord(HoodieRecord<?> record, Collector<O> out) throws Exce
     } else {
       location = getNewRecordLocation(partitionPath);
       this.context.setCurrentKey(recordKey);
-      if (isChangingRecords) {
-        updateIndexState(partitionPath, location);
-      }
+    }
+    // always refresh the index
+    if (isChangingRecords) {
+      updateIndexState(partitionPath, location);
```

Comment:
After this change, shall L189 be redundant?

Reply:
No, but the code can be promoted indeed.